src/storage/src/table/batch_table/storage_table.rs

```diff
@@ -455,9 +459,11 @@ impl<S: StateStore, SD: ValueRowSerde> StorageTableInner<S, SD> {
             // Vnodes that are set and should be accessed.
             let vnodes = match vnode_hint {
                 // If `vnode_hint` is set, we can only access this single vnode.
-                Some(vnode) => Either::Left(std::iter::once(vnode)),
+                Some(vnode) if self.vnodes.is_set(vnode.to_index()) => {
+                    Either::Left(std::iter::once(vnode))
+                }
                 // Otherwise, we need to access all vnodes of this table.
-                None => Either::Right(self.vnodes.iter_vnodes()),
+                _ => Either::Right(self.vnodes.iter_vnodes()),
```

Comment:
If [compute_vnode goes this branch](https://github.com/risingwavelabs/risingwave/blob/fd5334d2bcd930f88e55dc4a1d8a56bc5a0c73a1/src/storage/src/table/mod.rs#L124), assertion doesn't happen.
So without this vnode check, if a singleton scanning is scheduled to multiple PUs, it may result duplicated rows (vnode_hint = DEFAULT_VNODE).
Although currently we've ensured this kind of query will only be scheduled to one PU, this change aims to be a safeguard.

Reply:
> this change aims to be a safeguard.

I don't think this is a good practice. 👀 If we intentionally schedule multiple scans for the same partition of a distributed table, then duplicated rows are still possible in the outputs after merging. In other words, it's the scheduler's responsibility for scheduling the tasks into correct parallel units and the table should not tolerate incorrect distribution.